File: DotNet/Tests/MessageFetcherTest.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Shastra.Hydra.Messaging;
using Shastra.Hydra.Messaging.MessageFetchers;
using Shastra.Hydra.Messaging.MessageIds;
using Shastra.Hydra.Messaging.Storage;
using Shastra.Hydra.Tests.Mocks;

namespace Shastra.Hydra.Tests
{
    [TestClass]
    public class MessageFetcherTest
    {
        private const string MessageSource = "MessageFetcherTest";
        private const string Alternating1 = "Alternating1";
        private const string Alternating2 = "Alternating2";
        private const int AlternatingMessageCount = 5;
        private static IStore _singleMessageStore;
        private static IStore _alternatingStore;

        [ClassInitialize]
        static public void Initialize(TestContext context)
        {
            // Set up stores
            _singleMessageStore = new MockStore("SingleMessageStore", "");
            var provider = new NearestServerProvider(new List<IStore> { _singleMessageStore });
            var service = new StdHydraService(provider);
            service.Send(new HydraMessage { Topic = "Test", Source = MessageSource, Data = "Test data" });

            _alternatingStore = new MockStore("AlternatingStore", "");
            provider = new NearestServerProvider(new List<IStore> { _alternatingStore });
            service = new StdHydraService(provider);
            for (int ii=0; ii < AlternatingMessageCount; ii++) {
                service.Send(new HydraMessage { Topic = Alternating1, Source = MessageSource, Data = string.Format("{0} message {1}", Alternating1, ii) });
                service.Send(new HydraMessage { Topic = Alternating2, Source = MessageSource, Data = string.Format("{0} message {1}", Alternating2, ii) });
            }
        }

        [TestMethod]
        public void TestSingleMessageByTopicFetcher()
        {
            Assert.AreEqual(_singleMessageStore.GetLastSeq(), 0, "LastSeq should be zero after sending one message");

            var fetcher = new HydraByTopicMessageFetcher("Test");
            var messages = fetcher.MessagesAfterIdUpToSeq(_singleMessageStore, MessageIdManager.Create(DateTime.UtcNow.AddHours(-1)), 1);
            Assert.AreEqual(1, messages.Count(), "Should receive one message for topic Test");

            fetcher = new HydraByTopicMessageFetcher("Other");
            messages = fetcher.MessagesAfterIdUpToSeq(_singleMessageStore, MessageIdManager.Create(DateTime.UtcNow.AddHours(-1)), 1);
            Assert.AreEqual(0, messages.Count(), "Should receive no messages for topic Other");
        }

        [TestMethod]
        public void TestAlternatingByTopicFetcher()
        {
            Assert.AreEqual(_alternatingStore.GetLastSeq(), AlternatingMessageCount * 2 - 1, "LastSeq should be nine after sending ten messages");

            var fetcher = new HydraByTopicMessageFetcher(Alternating1);
            // All Alternating1 messages
            var messages = fetcher.MessagesAfterIdUpToSeq(_alternatingStore, MessageIdManager.Create(DateTime.UtcNow.AddHours(-1)), AlternatingMessageCount * 2);
            Assert.AreEqual(AlternatingMessageCount, messages.Count(), string.Format("Should receive {0} messages for topic {1}", AlternatingMessageCount, Alternating1));

            // Initial messages up to a SeqId
            messages = fetcher.MessagesAfterIdUpToSeq(_alternatingStore, MessageIdManager.Create(DateTime.UtcNow.AddHours(-1)), 4);
            Assert.AreEqual(3, messages.Count(), string.Format("Should receive three messages for topic {0} up to SeqId 4", Alternating1));

            // All messages after the first
            var firstMessageId = messages.First().MessageId;
            messages = fetcher.MessagesAfterIdUpToSeq(_alternatingStore, firstMessageId, AlternatingMessageCount * 2);
            Assert.AreEqual(AlternatingMessageCount - 1, messages.Count(), string.Format("Should receive {0} messages for topic {1} after first MessageId", AlternatingMessageCount - 1, Alternating1));

            // Messages after the first, up to a SeqId
            messages = fetcher.MessagesAfterIdUpToSeq(_alternatingStore, firstMessageId, 4);
            Assert.AreEqual(2, messages.Count(), string.Format("Should receive two messages for topic {0} after first MessageId, up to SeqId 4", Alternating1));

            fetcher = new HydraByTopicMessageFetcher("Other");
            messages = fetcher.MessagesAfterIdUpToSeq(_alternatingStore, MessageIdManager.Create(DateTime.UtcNow.AddHours(-1)), 1);
            Assert.AreEqual(0, messages.Count(), "Should receive no messages for topic Other");
        }
    }
}

File: Java/hydra/src/test/java/uk/co/shastra/hydra/messaging/messagefetchers/MessageFetcherTest.java
package uk.co.shastra.hydra.messaging.messagefetchers;

import static org.junit.Assert.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;

import org.junit.BeforeClass;
import org.junit.Test;

import uk.co.shastra.hydra.messaging.HydraMessage;
import uk.co.shastra.hydra.messaging.StdHydraService;
import uk.co.shastra.hydra.messaging.HydraService;
import uk.co.shastra.hydra.messaging.messageids.MessageId;
import uk.co.shastra.hydra.messaging.messageids.MessageIdManager;
import uk.co.shastra.hydra.messaging.mocks.MockStore;
import uk.co.shastra.hydra.messaging.storage.NearestServerProvider;
import uk.co.shastra.hydra.messaging.storage.Provider;
import uk.co.shastra.hydra.messaging.storage.Store;

public class MessageFetcherTest {

    private static final String MessageSource = "MessageFetcherTest";
    private static final String Alternating1 = "Alternating1";
    private static final String Alternating2 = "Alternating2";
    private static final int AlternatingMessageCount = 5;
    private static Store singleMessageStore;
    private static Store alternatingStore;
    
	@BeforeClass
	public static void setUpBeforeClass() throws Exception {
        // Set up stores
        singleMessageStore = new MockStore("SingleMessageStore", "");
        Provider provider = new NearestServerProvider(new ArrayList<Store>(Arrays.asList(singleMessageStore)));
        HydraService service = new StdHydraService(provider);
        HydraMessage message = makeMessage("Test", "Test data");
        service.send(message);

        alternatingStore = new MockStore("AlternatingStore", "");
        provider = new NearestServerProvider(new ArrayList<Store>(Arrays.asList(alternatingStore)));
        service = new StdHydraService(provider);
        for (int ii=0; ii < AlternatingMessageCount; ii++) {
            service.send(makeMessage(Alternating1, String.format("%1$ss message %2$s", Alternating1, ii)));
            service.send(makeMessage(Alternating2, String.format("%1$ss message %2$s", Alternating2, ii)));
        }
	}


	@Test
	public void testSingleMessageByTopicFetcher() {
        assertEquals("LastSeq should be zero after sending one message", singleMessageStore.getLastSeq(), 0);

        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());
        cal.add(Calendar.HOUR_OF_DAY, -1);
        
        MessageFetcher<HydraMessage> fetcher = new HydraByTopicMessageFetcher("Test");
        Iterable<HydraMessage> messages = fetcher.messagesAfterIdUpToSeq(singleMessageStore, MessageIdManager.create(cal.getTime()), 1);
        assertEquals("Should receive one message for topic Test", 1, ((ArrayList<HydraMessage>)messages).size());

        fetcher = new HydraByTopicMessageFetcher("Other");

        messages = fetcher.messagesAfterIdUpToSeq(singleMessageStore, MessageIdManager.create(cal.getTime()), 1);
        assertEquals("Should receive no messages for topic Other", 0, ((ArrayList<HydraMessage>)messages).size());
	}
	
	@Test
    public void TestAlternatingByTopicFetcher()
    {
        assertEquals("LastSeq should be nine after sending ten messages", alternatingStore.getLastSeq(), AlternatingMessageCount * 2 - 1);

        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());
        cal.add(Calendar.HOUR_OF_DAY, -1);
        
        MessageFetcher<HydraMessage> fetcher = new HydraByTopicMessageFetcher(Alternating1);
        // All Alternating1 messages
        ArrayList<HydraMessage> messages = (ArrayList<HydraMessage>) fetcher.messagesAfterIdUpToSeq(alternatingStore, MessageIdManager.create(cal.getTime()), AlternatingMessageCount * 2);
        assertEquals(String.format("Should receive %1$s messages for topic %2$s", AlternatingMessageCount, Alternating1), AlternatingMessageCount, messages.size());

        // Initial messages up to a SeqId
        messages = (ArrayList<HydraMessage>) fetcher.messagesAfterIdUpToSeq(alternatingStore, MessageIdManager.create(cal.getTime()), 4);
        assertEquals(String.format("Should receive three messages for topic %1$s up to SeqId 4", Alternating1), 3, messages.size());

        // All messages after the first
        MessageId firstMessageId = messages.get(0).getMessageId();
        messages = (ArrayList<HydraMessage>) fetcher.messagesAfterIdUpToSeq(alternatingStore, firstMessageId, AlternatingMessageCount * 2);
        assertEquals(String.format("Should receive %1$s messages for topic %2$s after first MessageId", AlternatingMessageCount - 1, Alternating1), AlternatingMessageCount - 1, messages.size());

        // Messages after the first, up to a SeqId
        messages = (ArrayList<HydraMessage>) fetcher.messagesAfterIdUpToSeq(alternatingStore, firstMessageId, 4);
        assertEquals(String.format("Should receive two messages for topic %1$s after first MessageId, up to SeqId 4", Alternating1), 2, messages.size());

        fetcher = new HydraByTopicMessageFetcher("Other");
        messages = (ArrayList<HydraMessage>) fetcher.messagesAfterIdUpToSeq(alternatingStore, MessageIdManager.create(cal.getTime()), 1);
        assertEquals("Should receive no messages for topic Other", 0, messages.size());
    }

	private static HydraMessage makeMessage(String topic, String data) 
	{
		HydraMessage message = new HydraMessage();
        message.setTopic(topic);
        message.setSource(MessageSource);
        message.setData(data);
		return message;
	}
}
